modules/parent-join/src/main/java/org/elasticsearch/join/mapper/Joiner.java

```diff
@@ -0,0 +1,203 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Utility class to help build join queries and aggregations, based on a join_field
+ */
+public class Joiner {
+
+    /**
+     * Get the Joiner for this context, or {@code null} if none is configured
+     */
+    public static Joiner getJoiner(QueryShardContext context) {
+        return getJoiner(context::isFieldMapped, context::getFieldType);
+    }
+
+    /**
+     * Get the Joiner for this context, or {@code null} if none is configured
+     */
+    public static Joiner getJoiner(AggregationContext context) {
+        return getJoiner(context::isFieldMapped, context::getFieldType);
+    }
+
+    /**
+     * Get the Joiner for this context, or {@code null} if none is configured
+     */
+    public static Joiner getJoiner(Predicate<String> isMapped, Function<String, MappedFieldType> getFieldType) {
+        if (isMapped.test(MetaJoinFieldMapper.NAME) == false) {
+            return null;
+        }
+        MetaJoinFieldMapper.MetaJoinFieldType ft
+            = (MetaJoinFieldMapper.MetaJoinFieldType) getFieldType.apply(MetaJoinFieldMapper.NAME);
+        String joinField = ft.getJoinField();
+        if (isMapped.test(joinField) == false) {
+            return null;
+        }
+        ParentJoinFieldMapper.JoinFieldType jft =
+            (ParentJoinFieldMapper.JoinFieldType) getFieldType.apply(joinField);
+        return jft.getJoiner();
+    }
+
+    private final Map<String, Set<String>> parentsToChildren = new HashMap<>();
+    private final Map<String, String> childrenToParents = new HashMap<>();
+
+    private final String joinField;
+
+    /**
+     * Constructs a Joiner based on a join field and a set of relations
+     */
+    public Joiner(String joinField, List<Relations> relations) {
+        this.joinField = joinField;
+        for (Relations r : relations) {
+            for (String child : r.children) {
+                parentsToChildren.put(r.parent, r.children);
+                if (childrenToParents.containsKey(child)) {
+                    throw new IllegalArgumentException("[" + child + "] cannot have multiple parents");
+                }
+                childrenToParents.put(child, r.parent);
+            }
+        }
+    }
+
+    /**
+     * @return the join field for the index
+     */
+    public String getJoinField() {
+        return joinField;
+    }
+
+    /**
+     * @return a filter for documents of a specific join type
+     */
+    public Query filter(String relationType) {
+        return new TermQuery(new Term(joinField, relationType));
+    }
+
+    /**
+     * @return a filter for parent documents of a specific child type
+     */
+    public Query parentFilter(String childType) {
+        return new TermQuery(new Term(joinField, childrenToParents.get(childType)));
+    }
+
+    /**
+     * @return a filter for child documents of a specific parent type
+     */
+    public Query getChildrenFilter(String parentType) {
```

Comment:
nit: is there a reason why this method starts with get while the two above don't?